Sources/AWSSDKSwiftCore/Credential/GroupCredentialProvider.swift

```diff
@@ -0,0 +1,87 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the AWSSDKSwift open source project
+//
+// Copyright (c) 2017-2020 the AWSSDKSwift project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of AWSSDKSwift project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+import NIOConcurrencyHelpers
+
+/// get credentials from a list of possible credential providers
+class GroupCredentialProvider: CredentialProvider {
+    /// the provider chosen to supply credentials
+    var internalProvider: CredentialProvider? {
+        get {
+            self.lock.withLock {
+                _internalProvider
+            }
+        }
+    }
+    /// promise to find a credential provider
+    let startupPromise: EventLoopPromise<CredentialProvider>
+    
+    private let lock = Lock()
+    private var _internalProvider: CredentialProvider? = nil
+
+    init(providers: [CredentialProviderFactory]? = nil, context: CredentialProviderFactory.Context) {
+        #if os(Linux)
```

Comment:
I think the list of providers used is probably something the `AWSClient` should decide, so will make providers list non-optional and move the list construction out to `AWSClient.init`

Reply:
Will now move this to `CredentialProviderFactory.default()` when all providers are ready